core/src/apps/bitcoin/sign_tx/payment_request.py

```diff
@@ -0,0 +1,97 @@
+from micropython import const
+from typing import TYPE_CHECKING
+
+from storage import cache
+from trezor import wire
+from trezor.crypto.curve import secp256k1
+from trezor.crypto.hashlib import sha256
+from trezor.utils import HashWriter
+
+from apps.common import coininfo
+from apps.common.address_mac import check_address_mac
+from apps.common.keychain import Keychain
+
+from .. import writers
+
+if TYPE_CHECKING:
+    from trezor.messages import TxAckPaymentRequest, TxOutput
+
+_MEMO_TYPE_TEXT = const(1)
+_MEMO_TYPE_REFUND = const(2)
+_MEMO_TYPE_COIN_PURCHASE = const(3)
+
+
+class PaymentRequestVerifier:
+    if __debug__:
+        # secp256k1 public key of m/0h for "all all ... all" seed.
+        PUBLIC_KEY = b"\x03\x0f\xdf^(\x9bZ\xefSb\x90\x95:\xe8\x1c\xe6\x0e\x84\x1f\xf9V\xf3f\xac\x12?\xa6\x9d\xb3\xc7\x9f!\xb0"
+    else:
+        PUBLIC_KEY = b""
+
+    def __init__(
+        self, msg: TxAckPaymentRequest, coin: coininfo.CoinInfo, keychain: Keychain
+    ) -> None:
+        self.h_outputs = HashWriter(sha256())
+        self.amount = 0
+        self.expected_amount = msg.amount
+        self.signature = msg.signature
+        self.h_pr = HashWriter(sha256())
+
+        if msg.nonce:
+            nonce = bytes(msg.nonce)
+            if cache.get(cache.APP_COMMON_NONCE) != nonce:
+                raise wire.DataError("Invalid nonce in payment request.")
+            cache.delete(cache.APP_COMMON_NONCE)
+        else:
+            nonce = b""
+            if msg.memos:
+                wire.DataError("Missing nonce in payment request.")
+
+        writers.write_bytes_fixed(self.h_pr, b"SL\x00\x24", 4)
+        writers.write_bytes_prefixed(self.h_pr, nonce)
+        writers.write_bytes_prefixed(self.h_pr, msg.recipient_name.encode())
+        writers.write_compact_size(self.h_pr, len(msg.memos))
+        for m in msg.memos:
+            if m.text_memo is not None:
+                memo = m.text_memo
+                writers.write_uint32(self.h_pr, _MEMO_TYPE_TEXT)
+                writers.write_bytes_prefixed(self.h_pr, memo.text.encode())
+            elif m.refund_memo is not None:
+                memo = m.refund_memo
+                check_address_mac(memo.address, memo.mac, coin.slip44, keychain)
```

Comment:
one very small nit: either here or in the protobuf definition we should mention in a comment that this is using the SLIP44 ID of the same coin for which the payment request applies -- unlike the purchase memo, where the SLIP44 may (and usually will) differ

Reply:
https://github.com/trezor/trezor-firmware/pull/2069/files#diff-df31c4d0ad39cb3c36fb3d14a52b556c582b45fb1f5d199a49ce47c6304563e1R61